pkg/reconciler/apis/apiexportendpointslice/apiexportendpointslice_reconcile.go

```diff
@@ -0,0 +1,146 @@
+*/
+
+package apiexportendpointslice
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"path"
+
+	"github.com/kcp-dev/logicalcluster/v3"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog/v2"
+
+	virtualworkspacesoptions "github.com/kcp-dev/kcp/cmd/virtual-workspaces/options"
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	conditionsv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/apis/conditions/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/util/conditions"
+	"github.com/kcp-dev/kcp/pkg/logging"
+	apiexportbuilder "github.com/kcp-dev/kcp/pkg/virtual/apiexport/builder"
+)
+
+func (c *controller) reconcile(ctx context.Context, apiExportEndpointSlice *apisv1alpha1.APIExportEndpointSlice) error {
+
+	// TODO (fgiloux): When the information is available in the cache server
+	// check if at least one APIBinding is bound in the shard to the APIExport referenced by the APIExportEndpointSLice.
+	// If so, add the respective endpoint to the status.
+	// For now the unfiltered list is added.
+
+	// Get APIExport
+	apiExportPath := logicalcluster.NewPath(apiExportEndpointSlice.Spec.APIExport.Path)
+	if apiExportPath.Empty() {
+		apiExportPath = logicalcluster.From(apiExportEndpointSlice).Path()
+	}
+	apiExport, err := c.getAPIExport(logicalcluster.From(apiExportEndpointSlice), apiExportPath, apiExportEndpointSlice.Spec.APIExport.Name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Don't keep the endpoints if the APIExport has been deleted
+			apiExportEndpointSlice.Status.APIExportEndpoints = nil
+			conditions.MarkFalse(
+				apiExportEndpointSlice,
+				apisv1alpha1.APIExportValid,
+				apisv1alpha1.APIExportNotFoundReason,
+				conditionsv1alpha1.ConditionSeverityError,
+				"APIExport %s|%s not found",
+				apiExportPath,
+				apiExportEndpointSlice.Spec.APIExport.Name,
+			)
+			return nil
+		} else {
+			conditions.MarkFalse(
+				apiExportEndpointSlice,
+				apisv1alpha1.APIExportValid,
+				apisv1alpha1.InternalErrorReason,
+				conditionsv1alpha1.ConditionSeverityError,
+				"Error getting APIExport %s|%s",
+				apiExportPath,
+				apiExportEndpointSlice.Spec.APIExport.Name,
+			)
+			return err
+		}
+	}
+	conditions.MarkTrue(apiExportEndpointSlice, apisv1alpha1.APIExportValid)
+
+	if err = c.updateEndpoints(ctx, apiExportEndpointSlice, apiExport); err != nil {
+		conditions.MarkFalse(
+			apiExportEndpointSlice,
+			apisv1alpha1.APIExportEndpointSliceURLsReady,
+			apisv1alpha1.ErrorGeneratingURLsReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			err.Error(),
+		)
+		return err
+	}
+	conditions.MarkTrue(apiExportEndpointSlice, apisv1alpha1.APIExportEndpointSliceURLsReady)
+
+	return nil
+}
+
+func (c *controller) updateEndpoints(ctx context.Context,
+	apiExportEndpointSlice *apisv1alpha1.APIExportEndpointSlice,
+	apiExport *apisv1alpha1.APIExport) error {
+	logger := klog.FromContext(ctx)
+	shards, err := c.listShards()
+	if err != nil {
```

Comment:
listing from not synced lister won't give you an error but an empty list :(

Reply:
nope, you should waiting for `waitForOptionalSync` when you start your controller.

Reply:
added that. Thanks for pointing it out. It will however only work once your PR has merged and waitForOptionalSync has been fixed.